plugin-locationcomponent/src/main/java/com/mapbox/maps/plugin/locationcomponent/LocationCompassEngine.kt

```diff
@@ -0,0 +1,321 @@
+package com.mapbox.maps.plugin.locationcomponent
+
+import android.content.Context
+import android.hardware.Sensor
+import android.hardware.SensorEvent
+import android.hardware.SensorEventListener
+import android.hardware.SensorManager
+import android.os.SystemClock
+import android.view.Surface
+import android.view.WindowManager
+import com.mapbox.common.Logger
+
+/**
+ * This class handles compass events and tracking the current device heading.
+ */
+internal class LocationCompassEngine(context: Context) : SensorEventListener {
+  private val windowManager: WindowManager =
+    context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
+  private val sensorManager: SensorManager =
+    context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
+  private val compassListeners = mutableListOf<CompassListener>()
+
+  // Not all devices have a compassSensor
+  private var compassSensor: Sensor? = null
+  private var gravitySensor: Sensor? = null
+  private var magneticFieldSensor: Sensor? = null
+
+  private val rotationMatrix = FloatArray(9)
+  private var rotationVectorValue: FloatArray? = null
+  private var lastHeading = 0f
+  private var compassUpdateNextTimestamp: Long = 0
+  private var gravityValues = FloatArray(3)
+  private var magneticValues = FloatArray(3)
+
+  init {
+    compassSensor = sensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR)
+    if (compassSensor == null) {
+      Logger.w(
+        TAG,
+        "Rotation vector sensor not supported on device, falling back to accelerometer and magnetic field."
+      )
+      gravitySensor = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
+      magneticFieldSensor = sensorManager.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD)
+    }
+  }
+
+  fun addCompassListener(compassListener: CompassListener) {
```

Comment:
is it possible to add one listener several times?
is yes, what will happen if it you try to remove it? it will be removed only once? if only once `unregisterSensorListeners` will not called

Reply:
That's possible, will add a check to remove duplicated listeners.